lib/bundler/definition.rb

```diff
@@ -709,12 +709,8 @@ def converge_locked_specs
       # and Gemfile.lock. If the Gemfile modified a dependency, but
       # the gem in the Gemfile.lock still satisfies it, this is fine
       # too.
-      locked_deps_hash = @locked_deps.inject({}) do |hsh, dep|
-        hsh[dep] = dep
-        hsh
-      end
       @dependencies.each do |dep|
-        locked_dep = locked_deps_hash[dep]
```

Comment:
Actually, it seems the test failure is related to the method `dependencies_for_source_changed?`.

`locked_deps_for_source = @locked_deps.select {|_, dep| dep.source == locked_source }` fails, but `locked_deps_for_source = @locked_deps.values.select {|dep| dep.source == locked_source }` does not. I failed to take into account that we need an array of sources after the fact to turn into a set.